snc.sh

```diff
@@ -227,31 +227,57 @@ function create_pvs() {
+     while ! ${OC} get nodes >/dev/null 2>&1; do
+         sleep 3
+     done
+     echo "API server is up"
+
+    # Force the rotation for secrets
+    ${OC} get secret -A -o json | ${JQ} -r '.items[] | select(.metadata.annotations."auth.openshift.io/certificate-not-after" | .!=null and fromdateiso8601<='$( date --date='+1year' +%s )') | "-n \(.metadata.namespace) \(.metadata.name)"' | \
+        xargs -n3 ${OC} patch secret -p='{"metadata": {"annotations": {"auth.openshift.io/certificate-not-after": null}}}'
 
-    # Wait for 5 min to make sure cluster is stable again.
+    # Delete the current csr signer to get new request.
+    ${OC} delete secrets/csr-signer-signer secrets/csr-signer -n openshift-kube-controller-manager-operator || true
+    
+    echo "Waiting for 5 mins to make sure cluster is stable again"
     sleep 300
 
-    # Remove the 24 hours certs and bootstrap kubeconfig
-    # this kubeconfig will be regenerated and new certs will be created in pki folder
-    # which will have 30 days validity.
-    ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo rm -fr /var/lib/kubelet/pki
-    ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo rm -fr /var/lib/kubelet/kubeconfig
-    ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo systemctl restart kubelet
+    # Retry 4 times to make sure kubelet certs are rotated correctly.
+    i=0
+    while [ $i -lt 4 ]; do
+        if ! ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo openssl x509 -checkend 2160000 -noout -in /var/lib/kubelet/pki/kubelet-client-current.pem; then
+            # Remove the 24 hours certs and bootstrap kubeconfig
+            # this kubeconfig will be regenerated and new certs will be created in pki folder
+            # which will have 30 days validity.
+            ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo rm -fr /var/lib/kubelet/pki
+            ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo rm -fr /var/lib/kubelet/kubeconfig
+            ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo systemctl restart kubelet
+
+	    # Wait until bootstrap csr request is generated with 5 min timeout
+	    timeout 300 bash -c -- "until ${OC} get csr | grep Pending; do echo 'Waiting for first CSR request.'; sleep 2; done"
+	    ${OC} get csr -ojsonpath='{.items[*].metadata.name}' | xargs ${OC} adm certificate approve
+            
+	    echo "Retry loop $i, wait for 60sec before starting next loop"
+            sleep 60
+	else
+            break
+        fi
+	i=$[$i+1]
+    done
 
-    # Wait until bootstrap csr request is generated with 5 min timeout
-    timeout 300 bash -c -- "until ${OC} get csr | grep Pending; do echo 'Waiting for first CSR request.'; sleep 2; done"
-    ${OC} get csr -ojsonpath='{.items[*].metadata.name}' | xargs ${OC} adm certificate approve
+    if ! ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- sudo openssl x509 -checkend 2160000 -noout -in /var/lib/kubelet/pki/kubelet-client-current.pem; then
```

Comment:
With the new hack, I guess we could do a more wide check: look at all certs in /etc/kubernetes and check their date. (clearly not for this PR, just a remark).

Reply:
sure, we can do that.